backend/inspirehep/indexer/tasks.py

```diff
@@ -144,7 +86,19 @@ def index_record(self, uuid, record_version=None, force_delete=None):
     else:
         InspireRecordIndexer().index(record)
 
+    papers_to_reindex = []
     if isinstance(record, LiteratureRecord):
-        process_references_for_record(record=record)
+        papers_to_reindex.extend(record.get_linked_papers_if_reference_changed())
+        if current_app.config.get("FEATURE_FLAG_ENABLE_SELF_CITATIONS"):
+            papers_to_reindex.extend(
+                record.get_all_connected_papers_of_modified_authors()
+            )
+            papers_to_reindex.extend(
+                record.get_all_connected_papers_of_modified_collaborations()
+            )
     if isinstance(record, AuthorsRecord):
-        process_author_papers_if_author_changed_name(record=record)
+        papers_to_reindex.extend(
+            record.get_linked_author_papers_if_author_changed_name()
+        )
```

Comment:
I would suggest going all the way and encapsulating this at the record level, so you can do here something like:
```python
records_to_reindex = record.get_linked_records_to_reindex()
```
and have a dummy implementation on `InspireRecord` that always returns an empty list and override it on `AuthorsRecord` and `LiteratureRecord`. WDYT?

Reply:
I prefer not to do this as there are already complains that we have to much logic inside InspireRecords opbjects.

Reply:
But personally I agree that it's logical to move it there.